From felipe.balbi@nokia.com  Wed Aug  6 15:15:59 2008
From: Felipe Balbi <felipe.balbi@nokia.com>
Date: Mon,  4 Aug 2008 13:53:52 +0300
Subject: usb: musb: fix hanging when rmmod gadget driver
To: linux-omap@vger.kernel.org, linux-usb@vger.kernel.org
Cc: Felipe Balbi <felipe.balbi@nokia.com>, Greg Kroah-Hartman <gregkh@suse.de>, David Brownell <david-b@pacbell.net>, Tony Lindgren <tony@atomide.com>
Message-ID: <1217847232-1653-1-git-send-email-felipe.balbi@nokia.com>


If we try to modprobe a second gadget driver before
rmmoding the first one, the reference for the first
gadget driver would get NULLed avoiding usb to change
gadget drivers later.

Cc: David Brownell <david-b@pacbell.net>
Cc: Tony Lindgren <tony@atomide.com>
Signed-off-by: Felipe Balbi <felipe.balbi@nokia.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/musb/musb_gadget.c |   16 +++++++---------
 1 file changed, 7 insertions(+), 9 deletions(-)

--- a/drivers/usb/musb/musb_gadget.c
+++ b/drivers/usb/musb/musb_gadget.c
@@ -1710,17 +1710,15 @@ int usb_gadget_register_driver(struct us
 
 	spin_unlock_irqrestore(&musb->lock, flags);
 
-	if (retval == 0)
+	if (retval == 0) {
 		retval = driver->bind(&musb->g);
-	if (retval != 0) {
-		DBG(3, "bind to driver %s failed --> %d\n",
-			driver->driver.name, retval);
-		musb->gadget_driver = NULL;
-		musb->g.dev.driver = NULL;
-	}
+		if (retval != 0) {
+			DBG(3, "bind to driver %s failed --> %d\n",
+					driver->driver.name, retval);
+			musb->gadget_driver = NULL;
+			musb->g.dev.driver = NULL;
+		}
 
-	/* start peripheral and/or OTG engines */
-	if (retval == 0) {
 		spin_lock_irqsave(&musb->lock, flags);
 
 		/* REVISIT always use otg_set_peripheral(), handling
